File: Principles of Programming Languages/PrologVsMercury/reverse.m
% --------------------------------------
% DEMO 2 – List Reversal (Mercury)
% --------------------------------------

:- module reverse.
:- interface.
:- import_module io.

:- pred main(io::di, io::uo) is det.

:- implementation.

:- import_module list, string, int.

% Recursive reverse
:- func my_reverse(list(int)) = list(int).
my_reverse([]) = [].
my_reverse([H | T]) = my_reverse(T) ++ [H].

% Tail-recursive reverse
:- func my_reverse_tr(list(int)) = list(int).
my_reverse_tr(L) = rev_acc(L, []).

:- func rev_acc(list(int), list(int)) = list(int).
rev_acc([], Acc) = Acc.
rev_acc([H | T], Acc) = rev_acc(T, [H | Acc]).

:- func show(list(int)) = string.
show(L) = string.join_list(", ", list.map(string.int_to_string, L)).

main(!IO) :-
    List = [1,2,3,4,5],
    R1 = my_reverse(List),
    R2 = my_reverse_tr(List),
    io.write_string("Original: [" ++ show(List) ++ "]\n", !IO),
    io.write_string("Reverse: [" ++ show(R1) ++ "]\n", !IO),
    io.write_string("Tail Rec Reverse: [" ++ show(R2) ++ "]\n", !IO).

File: Principles of Programming Languages/PrologVsMercury/reverse.pl
% --------------------------------------
% DEMO 2 – List Reversal (Prolog)
% --------------------------------------

:- initialization(main).

% Regular recursive reverse
my_reverse([], []).
my_reverse([H|T], R) :-
    my_reverse(T, RT),
    append(RT, [H], R).

% Tail-recursive reverse
my_reverse_tr(L, R) :-
    rev_acc(L, [], R).

rev_acc([], Acc, Acc).
rev_acc([H|T], Acc, R) :-
    rev_acc(T, [H|Acc], R).

main :-
    List = [1,2,3,4,5],
    my_reverse(List, R1),
    my_reverse_tr(List, R2),
    write('Original: '), write(List), nl,
    write('Reverse: '), write(R1), nl,
    write('Tail Rec Reverse: '), write(R2), nl,
    halt.
